router/core/graphql_handler.go

```diff
@@ -159,12 +169,43 @@ func (h *GraphQLHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 			return
 		}
 
-		if sharedPreparedPlan == nil {
-			requestLogger.Error("prepare plan is nil", zap.Error(err))
+		preparedPlan, ok = sharedPreparedPlan.(planWithMetaData)
+		if !ok {
+			requestLogger.Error("unexpected prepared plan type")
 			w.WriteHeader(http.StatusInternalServerError)
+			writeRequestErrors(graphql.RequestErrorsFromError(internalServerErrorErr), w, requestLogger)
 			return
 		}
-		preparedPlan = sharedPreparedPlan.(planWithExtractedVariables)
+	}
+
+	if h.gqlMetricsExporter != nil {
```

Comment:
How about moving this logic away from here and just calling one func?

Reply:
Done